README.md
# xltransport_py

Initial commit.
## On-line Documentation

More information on the **xltransport_py** project can be found
[here](https://jddixon.github.io/xltransport_py)


CHANGES
py/xltransport_py/CHANGES

v0.0.11
    2018-03-04
        * edit setup.py, tox config files                           SLOC  652
v0.0.10
    2017-12-22
        * pylinting, general code cleanup                           SLOC  650
v0.0.9
    2017-10-11
        * add projConfig.toml                                       SLOC  644
v0.0.8
    2017-09-19
        * implemented IPv4Address and related classes               SLOC  608
v0.0.7
    2017-08-10
        * rewrite setup.py, add tox files, edit cfg files           SLOC  542
v0.0.6
    2017-07-06
        * restructure for pytest                                    SLOC  395
v0.0.5
    2017-06-03
        * drop dist* from repos; plus minor edits                   SLOC  386
v0.0.4
    2017-02-13
        * add test_ipv4addr.py                                      SLOC  318
v0.0.3
    2017-02-12
        * add pylint, coverage, skeletal test                       SLOC  180
v0.0.2
    2017-02-11
        * add various configuration files
v0.0.1
    2017-02-10
        * initial commit                                            SLOC  126














test_basics.py
#!/usr/bin/env python3
# xltransport_py/test_basics.py

""" Currently just exercises test framework. """

#import hashlib
#import os
import time
import unittest

from rnglib import SimpleRNG
#from xltransport import *


class TestWhatever(unittest.TestCase):
    """ Currently just exercises test framework. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_whatever(self):
        """
        Tests nothing at all.
        """

        self.assertEqual(1, 1)

if __name__ == '__main__':
    unittest.main()


test_ipv4addr.py
#!/usr/bin/env python3
# xltransport_py/test_basics.py

""" Exercise code handling IPv4 addresses. """

#import hashlib
#import os
import time
import unittest

from rnglib import SimpleRNG
from xltransport import IPv4Address, AddressError


class TestIPv4Address(unittest.TestCase):
    """ Exercise code handling IPv4 addresses. """

    A1234 = bytearray([1, 2, 3, 4])
    A1235 = bytearray([1, 2, 3, 5])

    def setUp(self):
        self.rng = SimpleRNG(time.time())
        self.v4addr = None
        self.v4addr2 = None

    def tearDown(self):
        pass

    def test_bad_addr(self):
        try:
            v4addr = IPv4Address("1.2.3.4", -92)
            self.fail("didn't catch out of range port number -92")
        except AddressError:
            # success
            pass
        try:
            v4addr = IPv4Address("1.2.3.4", 65536)
            self.fail("didn't catch out of range port number 2^16")
        except AddressError:
            # success
            pass
        try:
            v4addr = IPv4Address("1.2.3.4.5", 5)
            self.fail("didn't catch 1.2.3.4.5")
        except AddressError:
            # success
            pass

        self.assertFalse(IPv4Address.is_valid_address(None))
        self.assertFalse(IPv4Address.is_valid_address("0"))
        self.assertFalse(IPv4Address.is_valid_address("0.0"))
        self.assertFalse(IPv4Address.is_valid_address("0.0.0"))
        self.assertTrue(IPv4Address.is_valid_address("0.0.0.0"))
        self.assertFalse(IPv4Address.is_valid_address("0.0.0.0.0"))

    def testV4AddrWithPort(self):
        v4addr = IPv4Address(self.A1234, 97)
        self.assertIsNotNone(v4addr)
        self.assertEqual(97, v4addr.getPort())
        iaddr = v4addr.get_inet_address()   # InetAddress
        byte_addr = v4addr.get_ip_address()
        self.assertEqual(4, byte_addr.length)
        addr_from_ia = iaddr.get_address()
        self.assertEqual(4, addr_from_ia.length)
        for ndx in range(4):
            self.assertEqual(byte_addr[ndx], addr_from_ia[ndx])
        v4addr2 = IPv4Address(byte_addr, 97)
        assertIsNotNone(v4addr2)
        self.assertEqual(v4addr, v4addr2)
        self.assertEqual("1.2.3.4:97", v4addr2.__str__())

    def test_equal(self):
        v4addr = IPv4Address(self.A1234, 52)
        self.assertTrue(v4addr.equals(v4addr))

        v4addr2 = IPv4Address(self.A1235, 52)    # different IP
        self.assertFalse(v4addr.equals(None))
        self.assertFalse(v4addr.equals(v4addr2))
        v4addr2 = IPv4Address(self.A1234, 53)    # different port
        self.assertFalse(v4addr.equals(v4addr2))
        v4addr2 = IPv4Address(self.A1234, 52)    # same IP and port
        self.assertTrue(v4addr.equals(v4addr2))

    def test_private_ips(self):
        # Web server running in 10/8
        v4addr = IPv4Address("10.0.0.1:80")
        addr = v4addr.get_ip_address()
        self.assertEqual(10, addr[0])
        self.assertEqual(0, addr[1])
        self.assertEqual(0, addr[2])
        self.assertEqual(1, addr[3])
        self.assertTrue(IPv4Address.is_private(addr))
        self.assertEqual(80, v4addr.getPort())

        # Web server running in 128.0/16
        v4addr = IPv4Address("128.0.12.121:8080")
        addr = v4addr.get_ip_address()
        self.assertEqual(-128, addr[0])
        self.assertEqual(0, addr[1])
        self.assertEqual(12, addr[2])
        self.assertEqual(121, addr[3])
        self.assertTrue(IPv4Address.is_private(addr))
        self.assertEqual(8080, v4addr.getPort())

        addr = IPv4Address("127.255.0.4:8080").get_ip_address()
        self.assertFalse(IPv4Address.is_private(addr))
        addr = IPv4Address("128.1.0.4:8080").get_ip_address()
        self.assertFalse(IPv4Address.is_private(addr))

        # Web server running in 172.16/12
        v4addr = IPv4Address("172.30.0.1:443")
        addr = v4addr.get_ip_address()
        self.assertEqual(-84, addr[0])
        self.assertEqual(30, addr[1])
        self.assertEqual(0, addr[2])
        self.assertEqual(1, addr[3])
        self.assertTrue(IPv4Address.is_private(addr))
        self.assertEqual(443, v4addr.getPort())

        addr = IPv4Address("172.15.0.1:443").get_ip_address()
        self.assertFalse(IPv4Address.is_private(addr))
        addr = IPv4Address("172.32.0.1:443").get_ip_address()
        self.assertFalse(IPv4Address.is_private(addr))

    def test_rfc3330(self):

        test = bytearray([0, 0, 0, 1])     # gets modified
        # 0/8
        self.assertTrue(IPv4Address.is_rfc3330(test))
        self.assertTrue(IPv4Address.is_rfc3330not_private(test))
        # 14/8
        test[0] = 14
        self.assertTrue(IPv4Address.is_rfc3330(test))
        # 24/8
        test[0] = 24
        self.assertTrue(IPv4Address.is_rfc3330(test))
        # 127/0
        test[0] = 127
        self.assertTrue(IPv4Address.is_rfc3330(test))

        # 169.254/16, link local
        test = IPv4Address("169.254.0.1:443").get_ip_address()
        self.assertEqual(-87, test[0])
        self.assertEqual(-2, test[1])
        self.assertTrue(IPv4Address.is_rfc3330(test))
        self.assertFalse(IPv4Address.is_private(test))

        # 192.0.2.0/24, test net
        test = IPv4Address("192.0.2.14:443").get_ip_address()
        self.assertEqual(-64, test[0])
        self.assertTrue(IPv4Address.is_rfc3330(test))
        self.assertFalse(IPv4Address.is_private(test))

        # 192.88.99.0/24, 6to4 relay anycast
        test = IPv4Address("192.88.99.14:443").get_ip_address()
        self.assertEqual(-64, test[0])
        self.assertTrue(IPv4Address.is_rfc3330(test))

        # 198.18/15, benchmark testing
        test = IPv4Address("198.18.99.14:443").get_ip_address()
        self.assertEqual(-58, test[0])
        self.assertTrue(IPv4Address.is_rfc3330(test))
        test = IPv4Address("198.19.99.14:443").get_ip_address()
        self.assertTrue(IPv4Address.is_rfc3330(test))

        # 224/4, multicast
        test = IPv4Address("224.18.99.14:443").get_ip_address()
        self.assertEqual(-32, test[0])
        self.assertTrue(IPv4Address.is_rfc3330(test))

        # 240/4, reserved
        v4addr = IPv4Address("240.18.99.14:443")
        test = v4addr.get_ip_address()
        self.assertEqual(-16, test[0])
        self.assertTrue(IPv4Address.is_rfc3330(test))
        self.assertEqual("240.18.99.14:443", v4addr.__str__())

    def test_constructor_with_host_name(self):
        v4addr = IPv4Address("www.xlattice.org", 80)
        test = v4addr.get_ip_address()
        # known to be globally routable ;-)
        self.assertFalse(IPv4Address.is_rfc3330(test))

if __name__ == '__main__':
    unittest.main()


tests/test_basics.py
#!/usr/bin/env python3
# xltransport_py/test_basics.py

""" Currently just exercises test framework. """

# import hashlib
# import os
import time
import unittest

from rnglib import SimpleRNG
# from xltransport import *


class TestWhatever(unittest.TestCase):
    """ Currently just exercises test framework. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_whatever(self):
        """
        Tests nothing at all.
        """

        self.assertEqual(1, 1)


if __name__ == '__main__':
    unittest.main()


src/xltransport/__init__.py
# xltransport/__init__.py

""" Transport package for XLattice python library. """

import socket

assert socket       # SUPPRESS WARNING      # XXXXXXX

__version__ = '0.0.11'
__version_date__ = '2018-03-04'


def check_port(candidate):
    """
    Verify that a number is a valid IP port: a non-negative 16-bit
    integer.
    """
    port = int(candidate)
    if port < 0 or port >= 64 * 1024:
        raise AddressError("not a valid IP port: ", port)
    return port


class XLTransportError(RuntimeError):
    """ General purpose exception for the package. """


class AddressError(XLTransportError):
    """ Exceptions relating to XLattice Addresses. """
    pass


class Address(object):
    """ An XLattice Address. """

    def __eq__(self, other):
        raise NotImplementedError

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        raise NotImplementedError


class IPAddress(Address):
    """ An ipv4 or ipv6 address. """

    def __init__(self, nbo_host, port):
        self._nbo_host = nbo_host       # packed binary, network byte order
        self._port = check_port(port)

    @property
    def nbo_host(self):
        """ Return the host IP address as a bytes-like object. """
        return self._nbo_host

    @property
    def port(self):
        return self._port


class IPv4Address(IPAddress):
    """ An ipv4 address (dotted quad) """

    def __init__(self, host, port=0):
        nbo_host = socket.inet_aton(host)   # 'host' MUST be dotted quad
        super().__init__(nbo_host, port)

        # XXX need checks; also need to resolve domain name if found

    def __str__(self):
        return "%d.%d.%d.%d:%d" % (
            self.nbo_host[0], self.nbo_host[1],
            self.nbo_host[2], self.nbo_host[3],
            self.port)

    def __eq__(self, other):
        if other is None or not isinstance(other, IPv4Address):
            return False
        return self._nbo_host == other.nbo_host and self._port == other.port

    # XXX def clone(self):

    @staticmethod
    def is_valid_address(host_nbo):
        """ Return whether a valid ipV4 address. """
        return host_nbo is not None and len(host_nbo) == 4

    @staticmethod
    def is_private(host):
        """
        Return whether an address is in private address space (and so
        not globally routeable).
        """
        return host is not None and (
            host[0] == 10 or  # 10/8
            host[0] == 128 and host[1] == 0 or      # 128.0/16
            host[0] == 172 and host[1] >= 16 and host[1] < 32)  # 172.16/12

    @staticmethod
    def is_rfc3330_not_private(host):
        """
        Return whether a remote address is routeable.

        Allows blocks which are reserved by IANA but subject to allocation.
        """
        return IPv4Address.is_valid_address(host) and (
            (host[0] == 0) or                       # 'this', 0/8
            (host[0] == 14) or                      # public data network
            (host[0] == 24) or                      # 24/8, cable TV
            (host[0] == 127) or                     # 127/8, loopback
            (host[0] == 169 and host[1] == 254) or  # 169.254/16, link local
            (host[0] == 192 and
             (host[1] == 0 and host[2] == 2) or     # 192.0.2.0/24, test net
             (host[1] == 88 and host[2] == 99)) or  # 6to4 relay anycast
            (host[0] == 198 and (
                host[1] == 18 or host[1] == 19)) or  # benchmark testing
            (host[0] >= 224 and host[0] < 240) or   # 224/4, multicast
            (host[0] >= 240 and host[1] <= 255))    # 240/4, reserved

    @staticmethod
    def is_rfc3330(host):
        """ Return whether an address is RFC 3330-compliant. """
        return IPv4Address.is_private(host) or \
            IPv4Address.is_rfc3330_not_private(host)
